File: reuse_python/src/athlete_base.py
"""
Athlete Reusability Project

This project demonstrates different ways to reuse classes in Python:
- Inheritance
- Composition
- Composition through interface implementation

Classes:
    Athlete: Represents an athlete with a specific specialty.
    Sprinter: Represents a sprinter, inheriting from Athlete.
    Thrower: Represents a thrower, utilizing Athlete class through composition.
    AthleteInterface: Interface defining the method for athlete specialty.
    AthleteBase: Class implementing the AthleteInterface.
    DistanceRunner: Represents a distance runner, using Athlete interface through composition.
    Jumper: Represents a jumper, using Athlete interface through composition.

Usage:
    Import the required classes and create instances of different types of athletes.
    Each athlete instance can display information about themselves using the `display()` method.


Author: Dac Linh Giao NGUYEN
Date: 2024

"""
from src.athlete_interface import AthleteInterface


class AthleteBase(AthleteInterface):
    """Class implementing the AthleteInterface."""

    def __init__(self, speciality):
        """
        Initializes an AthleteImpl object with a given specialty.

        Args:
            speciality (str): The specialty of the athlete.
        """
        self.speciality = speciality

    def my_speciality(self):
        """
        Returns the specialty of the athlete.

        Returns:
            str: The specialty of the athlete.
        """
        return f"My speciality is {self.speciality}."

File: reuse_js/src/athlete_base.js
/**
 * Represents an implementation of AthleteInterface.
 *
 * @author Dac Linh Giao NGUYEN
 * @version 0.1
 * @since 0.1
 */
const AthleteInterface = require("./athlete_interface");

class AthleteBase extends AthleteInterface {
    /**
     * Create an Athlete implementation.
     *
     * @param {string} speciality - The speciality of the athlete.
     */
    constructor(speciality) {
        super();
        this.speciality = speciality;
    }

    /**
     * Get the speciality of the athlete.
     *
     * @returns {string} The speciality of the athlete.
     */
    mySpeciality() {
        return `My speciality is ${this.speciality}.`;
    }
}